### DISCUSSION TITLE
Reproduce model?

### DISCUSSION THREAD

**Discussion by simonprovost** [autogluon/autogluon#1902] [Q&A] [answered]
Hi every1ne,

I was wondering let's say I got a champion model, LightGBMXT. I want to reproduce it with the LGB & Sci-kit libraries. What is the step to follow to do this, please?

Best wishes,

**Comment by Innixma** [comment, marked as answer]
That would be very complicated, and you would essentially have to reimplement large parts of AutoGluon. My suggestion is to run a debugger through the code logic to understand everything that occurs during fit.

**Reply by simonprovost** [reply]
Crystal clear @Innixma. Thanks for your prompt answer, that is pretty helpful.

Consequently, I presume that the only remaining step (in my case) is to obtain the model generated by AutoGluon and utilise it as-is. Is this not a bit "black box" way of acting in a sense? I believe that in another AutoML framework, say, Auto-Sklearn, for instance, the outputted model can be redone by-hand with a few investigation of the output logs. I suppose, nonetheless, that it was the author's choice to execute things that way in AutoGluon. In any case, the outcome of AutoGluon was astoundingly fascinating given the time searching! I will investigate with what to proceed with my use case. 

I was also wondering if your AutoGluon might be a good fit for healthcare research, considering that it produces models that are extremely difficult to reproduce manually. Experience has shown that medical practitioners seek interpretability above everything else. How might we consider employing AutoGluon for medical research if we cannot even explain what is happening in greater detail to AI/ML practitioners? I am perplexed.

**Reply by simonprovost** [reply]
You may also regard your response as the answer for the post; if you have time to answer this extra question with the same scope, please do so; otherwise, let us leave it as-is. Thanks for your wonderful support already ✅

**Reply by Innixma** [reply]
Let me provide a counter-point: Why is LightGBM more interpretable than a black box? Why would re-implementing the model by hand make it more interpretable?

If you think of the feature importance available in LightGBM, XGBoost, Random Forest, etc., you may want to read this blog: https://explained.ai/rf-importance/

Feature importances calculated from these models natively are highly misleading. AutoGluon uses permutation shuffling feature importance which is a black-box interpretability technique that is much more reliable and accurate than model-specific methods via `predictor.feature_importance`.

If you are thinking of SHAP, [AutoGluon has full SHAP support](https://auto.gluon.ai/stable/tutorials/tabular_prediction/tabular-indepth.html#interpretability-feature-importance).

So returning to your final question:

> How might we consider employing AutoGluon for medical research if we cannot even explain what is happening in greater detail to AI/ML practitioners?

1. Ask the same question but replace "AutoGluon" with "LightGBM"
2. Follow-up: "Rather than using the LightGBM library, I want to implement LightGBM by hand, please [tell me how to do this](https://github.com/microsoft/LightGBM/graphs/contributors), otherwise it is blackbox" - See the problem?
3. `predictor.feature_importance`
4. [AutoGluon has full SHAP support](https://auto.gluon.ai/stable/tutorials/tabular_prediction/tabular-indepth.html#interpretability-feature-importance)
5. [AutoGluon has experimental support for human interpretable rule-based models](https://auto.gluon.ai/stable/tutorials/tabular_prediction/tabular-interpretability.html).

> I believe that in another AutoML framework, say, Auto-Sklearn, for instance, the outputted model can be redone by-hand with a few investigation of the output logs.

It is doubtful that you would get the same result doing this unless you replicate the data preprocessing done in the AutoML system or you are in the fortunate and rare position to have perfectly preprocessed data to send to the model. If any preprocessing is done, the question of interpretability can be asked again: "How can I trust and understand this data preprocessing", as now the preprocessing is effectively a part of the model logic.

**Reply by simonprovost** [reply]
That is what I would call an excellent and accurate explanation! I did not aware about the experimental support for human interpretable rule-based models, for instance. Regarding SHAP, I saw it plainly. Regarding the feature importance, I was unaware that you were implementing it with greater precision and reliability. Thank you so much for all the information, it might definitely be a game-changer now! I will investigate further and study the remaining papers you published as I have only read one thus far.

Wonderful day!

**Reply by HendricksJudy** [reply]
Thanks for your explanation of the predictor.feature_importance, it is very useful to understand the mechanism of the Autogluon!
